File: DIP/Employee.py
class Employee:
    def __init__(self, name, id):
        self.name = name
        self.id = id


class Manager(Employee):
    def calculate_salary(self):
        # Manager-specific salary calculation logic
        base_salary = 5000
        bonus = 2000
        return base_salary + bonus


class Developer(Employee):
    def calculate_salary(self):
        # Developer-specific salary calculation logic
        base_salary = 4000
        bonus = 1000
        return base_salary + bonus


def calculate_total_salary(employees):
    total_salary = 0
    for employee in employees:
        total_salary += employee.calculate_salary()
    return total_salary


manager = Manager("John Doe", 1)
developer = Developer("Jane Smith", 2)
employees = [manager, developer]

total_salary = calculate_total_salary(employees)
print("Total Salary:", total_salary)

File: DIP/Employee.cpp
#include <iostream>
#include <vector>

class Employee {
public:
    virtual ~Employee() {}
    virtual double calculateSalary() = 0;
};

class Manager : public Employee {
public:
    double calculateSalary() override {
        // Manager-specific salary calculation logic
        double baseSalary = 5000;
        double bonus = 2000;
        return baseSalary + bonus;
    }
};

class Developer : public Employee {
public:
    double calculateSalary() override {
        // Developer-specific salary calculation logic
        double baseSalary = 4000;
        double bonus = 1000;
        return baseSalary + bonus;
    }
};

double calculateTotalSalary(const std::vector<Employee*>& employees) {
    double totalSalary = 0;
    for (Employee *employee : employees) {
        totalSalary += employee->calculateSalary();
    }
    return totalSalary;
}

int main() {
    Manager manager;
    Developer developer;
    std::vector<Employee*> employees = {&manager, &developer};

    double totalSalary = calculateTotalSalary(employees);
    std::cout << "Total Salary: " << totalSalary << std::endl;

    return 0;
}
